Firestore/Example/Tests/Integration/API/FIRCursorTests.mm

```diff
@@ -192,4 +192,131 @@ - (void)testCanBeUsedInDescendingQueries {
+                                            nanoseconds:timestamp.nanoseconds + 3 * us],
+    }
+  }];
+
+  FIRQuerySnapshot *querySnapshot = [self
+      readDocumentSetForRef:[[testCollection queryWhereField:@"timestamp"
+                                      isGreaterThanOrEqualTo:timestamp]
+                                queryWhereField:@"timestamp"
+                                     isLessThan:[FIRTimestamp
+                                                    timestampWithSeconds:timestamp.seconds
+                                                             nanoseconds:timestamp.nanoseconds +
+                                                                         3 * us]]];
+  NSMutableArray<NSString *> *actual = [NSMutableArray array];
+  [querySnapshot.documents enumerateObjectsUsingBlock:^(FIRDocumentSnapshot *_Nonnull doc,
+                                                        NSUInteger idx, BOOL *_Nonnull stop) {
+    [actual addObject:doc.data[@"k"]];
+  }];
+  XCTAssertEqualObjects(actual, (@[ @"b", @"c", @"d" ]));
+}
+
+- (void)testTimestampsAreTruncatedToMicroseconds {
+  FIRTimestamp *nanos = [FIRTimestamp timestampWithSeconds:0 nanoseconds:123456789];
+  FIRTimestamp *micros = [FIRTimestamp timestampWithSeconds:0 nanoseconds:123456000];
+  FIRTimestamp *millis = [FIRTimestamp timestampWithSeconds:0 nanoseconds:123000000];
+  FIRCollectionReference *testCollection = [self collectionRefWithDocuments:@{
+    @"a" : @{@"k" : @"a", @"timestamp" : nanos},
+  }];
+
+  NSMutableArray<NSString *> *actual = [NSMutableArray array];
+
+  FIRQuerySnapshot *querySnapshot =
+      [self readDocumentSetForRef:[testCollection queryWhereField:@"timestamp" isEqualTo:nanos]];
+  [querySnapshot.documents enumerateObjectsUsingBlock:^(FIRDocumentSnapshot *_Nonnull doc,
+                                                        NSUInteger idx, BOOL *_Nonnull stop) {
+    [actual addObject:doc.data[@"k"]];
+  }];
+  XCTAssertEqualObjects(actual, (@[ @"a" ]));
+  [actual removeAllObjects];
+
+  // Because Timestamp should have been truncated to microseconds, the microsecond timestamp
+  // should be considered equal to the nanosecond one.
+  querySnapshot =
+      [self readDocumentSetForRef:[testCollection queryWhereField:@"timestamp" isEqualTo:micros]];
+  [querySnapshot.documents enumerateObjectsUsingBlock:^(FIRDocumentSnapshot *_Nonnull doc,
+                                                        NSUInteger idx, BOOL *_Nonnull stop) {
+    [actual addObject:doc.data[@"k"]];
+  }];
+  XCTAssertEqualObjects(actual, (@[ @"a" ]));
+  [actual removeAllObjects];
+
+  // The truncation is just to the microseconds, however, so the millisecond timestamp should be
+  // treated as different and thus the query should return no results.
+  querySnapshot =
+      [self readDocumentSetForRef:[testCollection queryWhereField:@"timestamp" isEqualTo:millis]];
+  [querySnapshot.documents enumerateObjectsUsingBlock:^(FIRDocumentSnapshot *_Nonnull doc,
```

Comment:
These tests are pretty noisy with a lot of boilerplate.

Since the keys and the value of @"k" are the same, consider just dropping the @"k" field and using the document ids for comparison purposes.

There's already a function for getting the IDs of all the documents: `FIRQuerySnapshotGetIDs`. Please find a way to share that rather than just hand rolling the expansion inline for each assertion.

Reply:
Thanks for the tips, implemented all of them and also added a helper function to wrap the micro/nanosecond stuff, much less boilerplate now.